File: examples/SingletonPattern.java
// ---------------------
// SingletonPattern.java
// ---------------------

class Eager {
    private static final Eager _only = new Eager();

    private Eager ()
        {}

    public static Eager only () {
        return _only;}

    public String f () {
        return "Eager.f()";}}

class Lazy1 {
    private static Lazy1 _only;

    private Lazy1 ()
        {}

    public static Lazy1 only () {
        if (Lazy1._only == null)
            Lazy1._only = new Lazy1();
        return Lazy1._only;}

    public String f () {
        return "Lazy1.f()";}}

class Lazy2 {
    private Lazy2 ()
        {}

    private static class Holder {
        private static final Lazy2 _only = new Lazy2();}

    public static Lazy2 only () {
        return Holder._only;}

    public String f () {
        return "Lazy2.f()";}}

public final class SingletonPattern {
    public static void test () {
    	assert(Eager.only()     == Eager.only());
    	assert(Eager.only().f() == "Eager.f()");

    	assert(Lazy1.only()     == Lazy1.only());
    	assert(Lazy1.only().f() == "Lazy1.f()");

    	assert(Lazy2.only()     == Lazy2.only());
    	assert(Lazy2.only().f() == "Lazy2.f()");}

    public static void main (String[] args) {
        System.out.println("SingletonPattern.java");
        test();
        System.out.println("Done.");}}

File: examples/SingletonPattern.py
#!/usr/bin/env python3

# pylint: disable = bad-whitespace
# pylint: disable = invalid-name
# pylint: disable = missing-docstring
# pylint: disable = too-few-public-methods

# -------------------
# SingletonPattern.py
# -------------------

class A :
    pass

def test1 () -> None :
    x = A()
    y = A()
    assert x != y

def singleton (c) :
    x = c()
    return lambda : x

@singleton
class B :
    pass

def test2 () -> None :
    x = B()
    y = B()
    assert x == y

if __name__ == "__main__" : # pragma: no cover
    print("SingletonPattern.py")
    test1()
    test2()
    print("Done.")
